### PR TITLE
keymap Greenshadowmaker for ergodash

### PR DESCRIPTION
**Pull request by GreenShadowMaker**
## Description  

keymap Greenshadowmaker for ergodash

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

N/A

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/config.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_USER_H
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/config.h]: If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`? 

It's simpler, and less prone to user error, so if you could make this switch? 

```suggestion
#pragma once
```

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/config.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#define CONFIG_USER_H
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/config.h]:
```suggestion
```

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/config.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#define USE_SERIAL
+// #define USE_I2C
+
+/* Select hand configuration */
+
+#define MASTER_LEFT
+// #define MASTER_RIGHT
+// #define EE_HANDS
+
+#endif
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/config.h]:
```suggestion
```

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/config.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_USER_H
+#define CONFIG_USER_H
+
+#include "../../config.h"
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/config.h]:
```suggestion
```

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c [hunk 5]
```diff
@@ -0,0 +1,156 @@
+#include QMK_KEYBOARD_H
+#include "action_layer.h"
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: Could you remove all of these additional includes?  

```suggestion
```

They get included automatically, at various points in the built process, so they are redundant and not necessary here.

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+#include "eeconfig.h"
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: Could you remove all of these additional includes?  

```suggestion
```

They get included automatically, at various points in the built process, so they are redundant and not necessary here.

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+enum custom_keycodes {
+  QWERTY = SAFE_RANGE,
+  LOWER,
+  RAISE,
+  ADJUST,
+};
+
+// Fillers to make layering more clear
+#define _______ KC_TRNS
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: These defines aren't necessary, as these are actually defined in the core.  You can safely remove them. 

https://github.com/qmk/qmk_firmware/blob/a0309db983150fd0197eb620a6ba552d90d29f93/quantum/quantum_keycodes.h#L30-L31


```suggestion
```

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+#define XXXXXXX KC_NO
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: These defines aren't necessary, as these are actually defined in the core.  You can safely remove them. 

https://github.com/qmk/qmk_firmware/blob/a0309db983150fd0197eb620a6ba552d90d29f93/quantum/quantum_keycodes.h#L30-L31


```suggestion
```

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+
+void persistent_default_layer_set(uint16_t default_layer) {
+  eeconfig_update_default_layer(default_layer);
+  default_layer_set(default_layer);
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+    case QWERTY:
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: If you are only have the one default layer, you should probably remove this.

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+      RESET, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                        XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,  \
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, RGB_TOG, BL_TOGG, XXXXXXX,                        XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, \
+      RAISE, XXXXXXX, RGB_HUI, RGB_SAI, RGB_VAI,  BL_INC, XXXXXXX,                        XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, \
+    EEP_RST, XXXXXXX, RGB_HUD, RGB_SAD, RGB_VAD,  BL_DEC,           XXXXXXX,     XXXXXXX,          XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, KC_RSFT, \
+    XXXXXXX, XXXXXXX, XXXXXXX,XXXXXXX,             LOWER,   KC_SPC ,XXXXXXX,     XXXXXXX,  KC_SPC,   RAISE,          XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX   \
+  )
+};
+
+
+void persistent_default_layer_set(uint16_t default_layer) {
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: Could you remove this function?  

It's been depreciated in favor of the `set_single_persistent_default_layer` function (which is a core function).

https://docs.qmk.fm/#/ref_functions?id=setting-the-persistent-default-layer

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/rules.mk [hunk 11]
```diff
@@ -0,0 +1,4 @@
+BACKLIGHT_ENABLE = yes
+RGBLIGHT_ENABLE = yes
+USE_I2C = no
```
Comment (drashna) [inline keyboards/ergodash/rev2/keymaps/greenshadowmaker/rules.mk]: This isn't actually used, and can be removed.

```suggestion
```

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c (hunk 10 above)
Comment (GreenShadowMaker) [reply keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: Removing, Thanks!  Didn't lookup what it was there for, I see now something I don't need/use  though may look at if I do make a different base map

File: keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c (hunk 9 above)
Comment (GreenShadowMaker) [reply keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: Just the QWERTY one?  I tried taking out the entire 

bool process_record_user(

But Layer switch didn't work after

Comment (GreenShadowMaker) [reply keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: I haven't figured out where the tri_layer magic all comes from

Comment (drashna) [reply keyboards/ergodash/rev2/keymaps/greenshadowmaker/keymap.c]: it comes from quantum.h, actually. 

And yeah, without the RAISE and LOWER keycodes, the layers won't work. 

https://docs.qmk.fm/#/ref_functions?id=olkb-tri-layers

**Comment by drashna** [comment]
Thanks!

**Pull request merged**